[validation] Fix bug on object type with "value" field + "list" option

rexxars opened the pull request:
<!-- Thank you for contributing to Sanity. Please read our [Code of Conduct](https://github.com/sanity-io/sanity/blob/next/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/sanity-io/sanity/blob/next/CONTRIBUTING.md) before submitting a PR.

To help us review your Pull Request, please make sure you follow the steps and guidelines below.

It's OK to open a Pull Request to start a discussion/ask for help, but it should then be created as a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/). 

Do not delete the instructional comments. -->

**Type of change (check at least one)**

- [x]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Documentation (fix or update to documentation)
- [ ]  Maintenance
- [ ]  Other, please describe:

**Does this change require a documentation update? (Check one)**

- [ ]  Yes
- [x]  No

**Current behavior**

Note: This is quite an edge-case, but be patient with me:

If you...
1. Have an object type that defines the field `value`
2. Add the (non-standard) `options.list` configuration (generally seen in `string` and `number` types) on the field/type
3. Set a (valid) value on the field 

... when the validation inference runs on the schema types, it will incorrectly grab the value of the `value` field, instead of the object itself. When comparing the value of the object field (`{value: 'foo', other: 'props'}`) against the "allowed" values, it will never find a match, since it's comparing the nested value of the `value` prop instead of the root object.

This results in a "Value does not match allowed values" error.

This particular case was uncovered by the [color-list](https://github.com/KimPaow/sanity-color-list) plugin.

**Description**

This PR changes the logic of the `extractValueFromListOption` method in the validation module to look for a defined `value` field on the object. If defined, it will use the root-level object as the value - otherwise it will assume you wanted the nested `value`.

While this technically is an edge case, I figured it's not too far-fetched that we will support an object select in the same way we support string/number selects, so this will come in handy in that case. 

I also added an example input component mirroring the behavior of the color list plugin (but without the fancy UI logic), which lets us reproduce the issue.

**Note for release**

<!-- Please include a high level summary of the changes this PR introduce. The intended audience is both editors and developers. If it's introducing a new feature, remember to link to docs/blogpost, if it's a bugfix, please describe the bug in non-technical terms (e.g. how a user/developer may have experienced it).
For inspiration, check out the release notes from an earlier release: https://github.com/sanity-io/sanity/releases/tag/v0.142.0 -->

- Fix bug when object type has a `value` field and is paired with the (unsupported) `options.list` schema type option

**Checklist** 

- [x]  I have read the [Contributing Guidelines](https://github.com/sanity-io/sanity/blob/next/CONTRIBUTING.md)
- [ ]  The PR title includes a link to the relevant issue
- [x]  The PR title is appropriately formatted: `[some-package] PR title (#123)`
- [x]  The code is linted
- [x]  The test suite is passing
- [ ]  Corresponding changes to the documentation have been made

benedicteb reviewed (approved):
Nice work! 😀

Since I'm thinking a lot of about automatic tests these days - can you picture a way this could be tested automatically? Like a "test script" or something with the browser?

rexxars commented:
> Nice work! 😀
> 
> Since I'm thinking a lot of about automatic tests these days - can you picture a way this could be tested automatically? Like a "list of actions" you can do with the test-studio to make sure this does what it should?

Thanks! 

This is a pretty deep edge-case, but you _should_ be able to test it with the new example I added to the test-studio;

- Open a document of type `validationTest`, scroll down to the `objectList` field
- Select a value from the custom input component (from the dropdown (`<select>`))
- Ensure that no validation errors show up

Having said that, I added some test cases in unit-test land which I ensured failed before the bugfix and passes now. 

Will merge this based on previous review given it's even more thorough now - thanks for pushing us to do more tests ❤️
